### PR TITLE
UTF-8 parsing

### PR DESCRIPTION
**Pull request by ruancomelli**
Very much a work in progress PR, creating this for CI + coderabbit

## Summary by CodeRabbit

- **New Features**
	- Introduced a robust UTF-8 parsing and validation framework to enhance character encoding handling.
	- Added a command-line option parsing library for improved management of arguments.

- **Build System Changes**
	- Expanded build targets in the Makefile to include new UTF-8 related tests.
	- Streamlined the handling of compiler flags for a more efficient build process.
	- Enhanced documentation generation process through Doxygen integration.

- **Chores**
	- Updated `.gitignore` to exclude specific files and directories from version control.
	- Added a new `.clang-tidy` configuration for enhanced static analysis.
	- Integrated a Git submodule for improved documentation styling.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is an experiment review for experiment `mermaid_diagrams`
Run ID: `mermaid_diagrams/run_2024-09-20T14-16-16_v1-22-0-140-ga8d2a8b27-dirty`
The benchmark review for this pull request can be found at https://github.com/sourcery-ai-experiments/libcanth/pull/1.
This pull request was cloned from `https://github.com/imaami/libcanth/pull/2`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/sourcery-ai/core/pull/4607
- https://github.com/sourcery-ai/core/pull/4631
- https://github.com/sourcery-ai/core/pull/4647
    # CodeRabbit examples:
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/alexsnow348/insightface/pull/46
- https://github.com/alikuxac/utilities/pull/10
- https://github.com/AlphaDev87/timba-api/pull/49
- https://github.com/AngeloTadeucci/Maple2/pull/239
- https://github.com/AngeloTadeucci/Maple2.File/pull/36
- https://github.com/AngeloTadeucci/Maple2/pull/233
    # Examples where CodeRabbit does not generate diagrams
- https://github.com/baptisteArno/typebot.io/pull/1778
- https://github.com/btipling/foundations/pull/33
- https://github.com/btipling/foundations/pull/31
- https://github.com/chintu-777/jaeger/pull/1
- https://github.com/coji/remix-docs-ja/pull/55
- https://github.com/DaveMBush/SmartNgRX/pull/622
- https://github.com/DaveMBush/SmartNgRX/pull/481
- https://github.com/dkittle/party-connections/pull/6
- https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
- https://github.com/imaami/libcanth/pull/2
  # Requested by Tim
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/4DNucleome/PartSeg/pull/1183
- https://github.com/ably/ably-cocoa/pull/1973
- https://github.com/ably-labs/ably-chat-swift/pull/54
- https://github.com/ably-labs/ably-chat-swift/pull/35

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4607
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/338
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4631
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/339
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4647
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/340
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/2
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/1
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/2
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/3
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/2
- dataset_pull_request: https://github.com/alikuxac/utilities/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/utilities/pull/2
- dataset_pull_request: https://github.com/AlphaDev87/timba-api/pull/49
  review_pull_request: https://github.com/sourcery-ai-experiments/timba-api/pull/2
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/239
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/3
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2.File/pull/36
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2.File/pull/2
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/233
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/4
- dataset_pull_request: https://github.com/baptisteArno/typebot.io/pull/1778
  review_pull_request: https://github.com/sourcery-ai-experiments/typebot.io/pull/1
- dataset_pull_request: https://github.com/btipling/foundations/pull/33
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/1
- dataset_pull_request: https://github.com/btipling/foundations/pull/31
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/2
- dataset_pull_request: https://github.com/chintu-777/jaeger/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/jaeger/pull/1
- dataset_pull_request: https://github.com/coji/remix-docs-ja/pull/55
  review_pull_request: https://github.com/sourcery-ai-experiments/remix-docs-ja/pull/1
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/622
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/1
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/481
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/2
- dataset_pull_request: https://github.com/dkittle/party-connections/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/party-connections/pull/1
- dataset_pull_request: https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/onstudy-backend/pull/1
- dataset_pull_request: https://github.com/imaami/libcanth/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/libcanth/pull/1
- dataset_pull_request: https://github.com/4DNucleome/PartSeg/pull/1183
  review_pull_request: https://github.com/sourcery-ai-experiments/PartSeg/pull/2
- dataset_pull_request: https://github.com/ably/ably-cocoa/pull/1973
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-cocoa/pull/1
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/54
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/1
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/35
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/2

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request introduces a robust UTF-8 parsing and validation framework, along with command-line option parsing capabilities. The changes include new source files for UTF-8 handling and option parsing, updates to the build system, and the addition of a test file for UTF-8 functionality. The implementation focuses on efficient state machine-based UTF-8 parsing and flexible command-line argument handling.

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Implemented a UTF-8 parsing and validation framework | <ul><li>Created a state machine-based UTF-8 parser</li><li>Implemented functions for parsing UTF-8 code points</li><li>Added utility functions for UTF-8 handling</li><li>Included error handling for invalid UTF-8 sequences</li></ul> | `src/utf8.c`<br/>`src/utf8.h` |
| Added a command-line option parsing library | <ul><li>Implemented a flexible option parsing system</li><li>Added support for boolean, string, and number arguments</li><li>Included help text generation functionality</li></ul> | `src/letopt.h`<br/>`src/letopt.c` |
| Created a test file for UTF-8 functionality | <ul><li>Implemented a command-line tool to inspect UTF-8 encoded strings</li><li>Added options for counting bytes, characters, and handling invalid sequences</li><li>Integrated the new UTF-8 parser and option parsing library</li></ul> | `src/test-utf8.c` |
| Updated the build system | <ul><li>Added new build targets for UTF-8 related tests</li><li>Updated compiler flags and source file lists</li></ul> | `src/Makefile` |
| Added documentation generation | <ul><li>Created a Makefile for generating documentation using Doxygen</li></ul> | `doc/Makefile` |

### Sequence Diagram

```mermaid
sequenceDiagram
    participant Caller
    participant Parser as UTF-8 Parser
    participant LUT as Lookup Table
    participant Cache

    Caller->>Parser: utf8_parse_next_code_point()
    Parser->>LUT: Check byte validity
    LUT-->>Parser: Return next state
    Parser->>Cache: Update parser cache
    alt Valid UTF-8 sequence
        Parser-->>Caller: Return next byte address
    else Invalid sequence
        Parser-->>Caller: Return error and invalid byte address
    end
```

---

<details>
<summary>Tips</summary>

- Trigger a new Sourcery review by commenting `@sourcery-ai review` on the pull request.
- Continue your discussion with Sourcery by replying directly to review comments.
- You can change your review settings at any time by accessing your [dashboard](https://app.sourcery.ai):
    - Enable or disable the Sourcery-generated pull request summary or reviewer's guide;
    - Change the review language;
- You can always [contact us](mailto:support@sourcery.ai) if you have any questions or feedback.

</details>

File: src/utf8.c [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+
+/**
+ * @brief Get a mask of state flags comprising the states allowed after `st8`.
+ *
+ * @param st8 A parser state enumeration.
+ * @return Mask of state flags representing the allowed states, or 0 if `st8`
+ *         is not a valid state enumeration.
+ */
+static const_inline uint16_t
+utf8_get_allowed_next_states (enum utf8_st8 st8)
```
Comment (SourceryAI) [inline src/utf8.c:272]: **suggestion:** Consider adding more comments to explain the state machine transitions

The UTF-8 parser implementation is complex and optimized. Additional comments explaining the state transitions would greatly improve code readability and maintainability.

File: src/letopt.h [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+        letopt_get_string_arg (struct letopt_state  *state,  \
+                               char const          **dest)
+
+#endif // OPTIONS || INCLUDED_FROM_LETOPT_C_
+
+#ifdef OPTIONS
+#include "ligma.h"
+#include "util.h"
+
+#define letopt_error(opt) ((opt)->p.e)
```
Comment (SourceryAI) [inline src/letopt.h:72]: **suggestion:** Consider simplifying the macro-heavy approach in letopt.h

While macros can be powerful, they can also make the code harder to debug and maintain. Consider simplifying this approach or adding more explanatory comments if simplification is not feasible.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes - here's some feedback:

**Overall Comments**:
- Consider simplifying the UTF-8 parser state machine if possible, without losing functionality. The current implementation, while efficient, is quite complex and might be challenging to maintain.
- The extensive use of macros, especially in letopt.h, could potentially be reduced. Consider if some of these could be replaced with inline functions or constexpr in C++ for improved readability and debuggability.
- The error handling in the UTF-8 parser uses errno values. Consider implementing a more explicit error handling mechanism, perhaps using an enum for error codes specific to the UTF-8 parser.

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 2 issues found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/798e4b35-0204-4ea7-b3f7-37bb26158c14?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**